### PR TITLE
Part 4 of RFC2906 - Add support for inheriting `readme`

### PR DESCRIPTION
**Pull request by Muscraft**
Tracking issue: #8415
RFC: rust-lang/rfcs#2906

[Part 1](https://github.com/rust-lang/cargo/pull/10497)
[Part 2](https://github.com/rust-lang/cargo/pull/10517)
[Part 3](https://github.com/rust-lang/cargo/pull/10538)

This PR focuses on adding support for inheriting `readme`:
- Added adjusting of a `readme` path when it is outside of the `package_root`
  - Copied from `license-file`'s implementation in `toml::prepare_for_publish()`
- Added copying of a `readme` file when it is outside of the `package_root` for publishing
  - Copied from `license-file`'s implementation in `cargo_package::build_ar_list()`
- Merged copying of a file outside of a `package_root` to reduce duplicated code and allow for other files in the future to be copied in a similar way

Remaining implementation work for the RFC
- Path dependencies infer version directive
- Lock workspace dependencies and warn when unused
- Optimizations, as needed
- Evaluate any new fields for being inheritable (e.g. `rust-version`)

### CODE REVIEW

File: src/cargo/ops/cargo_package.rs [hunk 1]
```diff
@@ -258,67 +258,91 @@ fn build_ar_list(
...
+    if let Some(readme_file) = &pkg.manifest().metadata().readme {
+        check_for_file_and_add("readme", readme_file, pkg, &mut result, ws)?
+    }
+
+    result.sort_unstable_by(|a, b| a.rel_path.cmp(&b.rel_path));
+
+    Ok(result)
+}
+
+fn check_for_file_and_add(
```
Comment (epage) [inline src/cargo/ops/cargo_package.rs:275]: Ideally this refactor would be done in a separate commit than the adding of readme support

File: src/cargo/ops/cargo_package.rs [hunk 2]
```diff
@@ -258,67 +258,91 @@ fn build_ar_list(
...
                 This may become a hard error in the future.",
-                license_path.display(),
+                label,
+                file_path.display(),
                 rel_msg,
+                label,
                 pkg.manifest_path().display()
             ))?;
         }
     }
```
Comment (epage) [inline src/cargo/ops/cargo_package.rs:344]: imo this shouldn't be in the shared function.  If we end up wanting to reuse it, it could be a function on its own.

File: src/cargo/util/toml/mod.rs [hunk 3]
```diff
@@ -1175,6 +1175,31 @@ impl TomlManifest {
...
+                                .unwrap()
+                                .to_str()
+                                .unwrap()
+                                .to_string(),
+                        )));
+                    }
+                }
+                StringOrBool::Bool(_) => {}
+            }
+        }
```
Comment (epage) [inline src/cargo/util/toml/mod.rs:1202]: Thoughts on this also being a function?

**Review by epage — COMMENTED** [review]
Thanks!

File: src/cargo/ops/cargo_package.rs (hunk 1 above)
Comment (Muscraft) [reply src/cargo/ops/cargo_package.rs:275]: I can move it to a separate commit if we keep it

File: src/cargo/ops/cargo_package.rs (hunk 2 above)
Comment (Muscraft) [reply src/cargo/ops/cargo_package.rs:344]: I agree, I'm just not sure where to have the error message since we check if the file exists inside of it. We could maybe change the function signature to take in a `warn_not_exist: bool` and remove `if label != "readme"`

File: src/cargo/ops/cargo_package.rs (hunk 1 above)
Comment (epage) [reply src/cargo/ops/cargo_package.rs:275]: Yes, I'm liking the refactor.  We can make it the first commit of this PR.

File: src/cargo/util/toml/mod.rs (hunk 3 above)
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1202]: We can if you think its better, I didn't since `license-file` is a `String` and `readme` is a `StringOrBool`

File: src/cargo/ops/cargo_package.rs (hunk 2 above)
Comment (epage) [reply src/cargo/ops/cargo_package.rs:344]: Adding bools to a function like that is suspiciousm, that we might be making a god-function.

One idea is if we keep the joining / existence check out of the function, then we should be able to make this work.

Comment (Muscraft) [reply src/cargo/ops/cargo_package.rs]: I did just that, let me know what you think.

**Review by epage — APPROVED** [review]
Looks great!
